Handle null arguments in the ContentCompositionListener

fritzmg opened the pull request:
Fixes #2787 ([thread](https://community.contao.org/de/showthread.php?80310-Internal-Server-Error-beim-Dr%C3%BCcken-auf-quot-Auswahl-%C3%A4ndern-quot-bei-der-Weiterleitungsseite&p=540163&viewfull=1#post540163)):

```
TypeError:
preg_replace(): Argument #3 ($subject) must be of type array|string, null given

  at vendor/contao/core-bundle/src/EventListener/DataContainer/ContentCompositionListener.php:107
  at preg_replace()
     (vendor/contao/core-bundle/src/EventListener/DataContainer/ContentCompositionListener.php:107)
  at Contao\CoreBundle\EventListener\DataContainer\ContentCompositionListener->renderPageArticlesOperation()
     (vendor/contao/core-bundle/src/Resources/contao/classes/DataContainer.php:804)
  at Contao\DataContainer->generateButtons()
     (vendor/contao/core-bundle/src/Resources/contao/drivers/DC_Table.php:4028)
  at Contao\DC_Table->generateTree()
     (vendor/contao/core-bundle/src/Resources/contao/drivers/DC_Table.php:3628)
  at Contao\DC_Table->treeView()
     (vendor/contao/core-bundle/src/Resources/contao/drivers/DC_Table.php:353)
  at Contao\DC_Table->showAll()
     (vendor/contao/core-bundle/src/Resources/contao/classes/Backend.php:644)
  at Contao\Backend->getBackendModule()
     (vendor/contao/core-bundle/src/Resources/contao/controllers/BackendMain.php:167)
  at Contao\BackendMain->run()
     (vendor/contao/core-bundle/src/Controller/BackendController.php:48)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor/symfony/http-kernel/HttpKernel.php:158)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (vendor/symfony/http-kernel/HttpKernel.php:80)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (vendor/symfony/http-kernel/Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle()
     (web/index.php:31)
```

Since `$icon` can specifically be null. Not sure why it would be though in this specific instance, but apparently it can happen.

fritzmg committed:
check if icon is null

leofeyer commented:
The PR is based on the wrong branch.

m-vo commented:
In the picker `$href` and `$icon` is null every time. If `!$this->pageRegistry->supportsContentComposition($pageModel) || !$this->hasArticlesInLayout($pageModel)` evaluates to `false` an `a` tag without inner HTML is output. We should IMHO always return early in this case.

The original issue probably appears because the mentioned condition evaluates to `true` (legacy routing mode or something?).

fritzmg commented:
> We should IMHO always return early in this case.

No, because you could still style the `<a>` otherwise.

// sorry, I meant we cannot return early, if only `$icon` is `null`. But we could if both `$href` and `$icon` are `null`, yes.

m-vo commented:
I can reproduce #2787 as well with PHP8.0.

If you have a page that has no layout assigned `$this->hasArticlesInLayout($pageModel)` returns `false` which is why the failing code is executed. And `icon`/`href` can be `null` according to the DCA definition. 

This must IMHO also be taken into account in line 112 (`$href = null`) as well as line 114 (`$icon = null`).

Btw.: If both are `null` we could imo shortcut execution but maybe someone relies on the anchor with a title: `<a href="" title="foobar"></a>` that is still output. Giving this a second thought, I'd leave this out and focus on the PHP8 problem first.

fritzmg commented:
Line 114 is ok, because there are no strict types. And `Image::getHtml` returns early with an empty string, if `$src` is falsey. Line 112 is ok too because concatenating a string together with `null` is allowed: https://3v4l.org/VCCr2

m-vo commented:
> Line 114 is ok, because there are no strict types. And Image::getHtml returns early with an empty string

While true, at least this one looks a bit fragile to me. 🙃

fritzmg committed:
return empty string if href and icon are missing

fritzmg commented:
I decided to include the return for when both the `$href` and `$icon` is `null` in this PR as well. I don't think it ever makes sense to output this HTML in the picker:

```html
<a href="contao?do=page&amp;popup=1&amp;picker=…&amp;pn=2&amp;rt=…&amp;ref=…" title=""></a>
```

This would be the link for the articles of a page within the page section of the picker - which should not be present anyway, but it's currently only not visible because it has not text or image content.

leofeyer commented on core-bundle/src/EventListener/DataContainer/ContentCompositionListener.php:
```diff
@@ -95,7 +95,7 @@ public function __construct(ContaoFramework $framework, Security $security, Page
     public function renderPageArticlesOperation(array $row, ?string $href, string $label, string $title, ?string $icon): string
     {
-        if (!$this->security->isGranted(ContaoCorePermissions::USER_CAN_ACCESS_MODULE, 'article')) {
+        if (null === $href && null === $icon || !$this->security->isGranted(ContaoCorePermissions::USER_CAN_ACCESS_MODULE, 'article')) {
```
```suggestion
        if ((null === $href && null === $icon) || !$this->security->isGranted(ContaoCorePermissions::USER_CAN_ACCESS_MODULE, 'article')) {
```

<img width="905" alt="" src="https://user-images.githubusercontent.com/1192057/109843615-8aeb4d00-7c4b-11eb-9d16-859f4e76461e.png">

leofeyer committed:
Merge branch '4.11' into fix-content-composition-listener-if-icon-is-null

m-vo committed:
Correctly merge image size _defaults (see #2783)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | -
| Docs PR or issue | -

Currently defining `formats` under the image size `_defaults` is broken. This is because the processed config values of the individual size definitions already contain an empty `formats` array which then wins in  `array_merge`. This PR fixes this problem by giving the values under `_defaults` precedence over empty arrays.

Unfortunately this case was missing from the tests and went unnoticed (added now)…

Commits
-------

b7b7b601 merge _defaults with precedence over empty arrays
990be331 add reference
af36ec25 Merge branch '4.11' into bugfix/image-size-defaults-empty-arrays

m-vo committed:
Suggest using the contao-setup binary with @php prefix (see #2796)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2776 
| Docs PR or issue | -

related https://github.com/contao/managed-edition/pull/57

Commits
-------

b1bd1850 suggest using the prefixed contao-setup binary

m-vo committed:
fix accessing undefined variable (see #2806)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #...
| Docs PR or issue | contao/docs#...

Sigh, another one for PHP8.

Fixes 
```
Warning: Undefined array key "breadcrumb"
```
after clicking on the articles selection in the page tree because the `sorting` array does not contain `breadcrumb`.

Commits
-------

36dcf5e6 fix accessing undefined variable

aschempp committed:
Correctly sort pages if the URL suffix is empty (see #2784)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2775

Apparently, `substr('foo', 0, -0)` is empty 🙈

Commits
-------

502a0e03 Correctly sort pages on empty URL suffix
aa5fe3b3 Merge branch '4.11' into bugfix/folderurl

m-vo committed:
Fix accessing Model\Collection instead of Model in ModuleFaqPage (see #2788)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2786 
| Docs PR or issue | -

`$figureBuilder->setMetadata($objFaq->getOverwriteMetadata())` was failing because `$objFaq` is a `Model\Collection` instead of a `Model`. To prevent this rather hard to spot error in the future, I adjusted how we're iterating over the results instead.

Commits
-------

ff2f8b9d iterate over FaqModels instead of operating on Model\Collection
c2934f0c Merge branch '4.11' into bugfix/module-faq-model-collection

m-vo committed:
Handle another illegal array access in the tl_page DCA (see #2794)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes -
| Docs PR or issue | -

This fixes `Warning: Undefined array key "root"` coming from `DCTable` line 3669. This also hardens `tl_page#pastePage()` that will receive an incomplete `$row` by the call in line 3678 and would otherwise produce `Warning: Undefined array key "pid"`.

https://github.com/contao/contao/blob/1873bcc9ea728a08ca8755f6e7251d999d7b316b/core-bundle/src/Resources/contao/drivers/DC_Table.php#L3669-L3690

Commits
-------

7e5b39f2 handle illegal array access
5ff8b5f4 simplify expression

Co-authored-by: Leo Feyer <github@contao.org>

fritzmg committed:
Fix the logout handler in Symfony 5 (see #2818)

Description
-----------

This PR fixes the following error in Contao **4.11.0**:

```
TypeError:
Argument 2 passed to Contao\CoreBundle\Security\Logout\LogoutHandler::logout() must be an instance of Symfony\Component\HttpFoundation\Response, null given, called in vendor\symfony\security-bundle\Security\LegacyLogoutHandlerListener.php on line 42

  at vendor\contao\contao\core-bundle\src\Security\Logout\LogoutHandler.php:50
  at Contao\CoreBundle\Security\Logout\LogoutHandler->logout()
     (vendor\symfony\security-bundle\Security\LegacyLogoutHandlerListener.php:42)
  at Symfony\Bundle\SecurityBundle\Security\LegacyLogoutHandlerListener->onLogout()
     (vendor\symfony\event-dispatcher\EventDispatcher.php:270)
  at Symfony\Component\EventDispatcher\EventDispatcher::Symfony\Component\EventDispatcher\{closure}()
     (vendor\symfony\event-dispatcher\EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
     (vendor\symfony\event-dispatcher\EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
     (vendor\symfony\security-http\Firewall\LogoutListener.php:122)
  at Symfony\Component\Security\Http\Firewall\LogoutListener->authenticate()
     (vendor\symfony\security-bundle\Debug\WrappedLazyListener.php:49)
  at Symfony\Bundle\SecurityBundle\Debug\WrappedLazyListener->authenticate()
     (vendor\symfony\security-http\Firewall\AbstractListener.php:26)
  at Symfony\Component\Security\Http\Firewall\AbstractListener->__invoke()
     (vendor\symfony\security-bundle\Debug\TraceableFirewallListener.php:62)
  at Symfony\Bundle\SecurityBundle\Debug\TraceableFirewallListener->callListeners()
     (vendor\symfony\security-http\Firewall.php:86)
  at Symfony\Component\Security\Http\Firewall->onKernelRequest()
     (vendor\symfony\event-dispatcher\Debug\WrappedListener.php:117)
  at Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke()
     (vendor\symfony\event-dispatcher\EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
     (vendor\symfony\event-dispatcher\EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
     (vendor\symfony\event-dispatcher\Debug\TraceableEventDispatcher.php:151)
  at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch()
     (vendor\symfony\http-kernel\HttpKernel.php:133)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (vendor\symfony\http-kernel\HttpKernel.php:79)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (vendor\symfony\http-kernel\Kernel.php:195)
  at Symfony\Component\HttpKernel\Kernel->handle()
     (web\index.php:31)
  at require('web\\index.php')
     (web\app.php:4)           
```

**Reproduction**:

1. Create a login form, login page and member.
2. Create a logout page.
3. Log into the front end.
4. Use the link in the menu for the logout.

Apparently the `LogoutEvent` might not contain a `Response` object any more.

Commits
-------

aa5b2e8d fix argument error in Symfony 5
bd06335f Merge branch '4.11' into fix-logout-handler

fritzmg committed:
Update core-bundle/src/EventListener/DataContainer/ContentCompositionListener.php

Co-authored-by: Leo Feyer <github@contao.org>

leofeyer commented on core-bundle/src/EventListener/DataContainer/ContentCompositionListener.php:
```diff
@@ -104,7 +104,7 @@ public function renderPageArticlesOperation(array $row, ?string $href, string $l
 
         if (!$this->pageRegistry->supportsContentComposition($pageModel) || !$this->hasArticlesInLayout($pageModel)) {
-            return $this->image->getHtml(preg_replace('/\.svg$/i', '_.svg', $icon)).' ';
+            return null !== $icon ? $this->image->getHtml(preg_replace('/\.svg$/i', '_.svg', $icon)).' ' : '';
```
This check is not necessary, because if `$icon` is `null`, we are returning early in line 98, aren't we?

fritzmg replied:
Line 98 only returns if both `$href` _and_ `$icon` are `null`. Granted, usually none or both will be null in this particular case, but we should employ defensive programming anyway, should we not?

leofeyer replied:
You are right.

m-vo committed:
Fix an illegal array access in DC_Table when expanding the tree (see #2805)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2797 
| Docs PR or issue | -

I split the PR in two commits:
* one that *only* formats the code so that there is any chance of reviewing this:
https://github.com/contao/contao/commit/a1a16fea1c1cae125f129f03e38db1a3abf5a59e
* one that applies the fixes:
https://github.com/contao/contao/commit/9bcedb793b74062b4e4cfddc7a30765694ffd5fe

(If you want to, I can collapse the whole thing into a oneline-mega-expression again in a third one…)

Commits
-------

a1a16fea format call
9bcedb79 fix illegal array access
2f22b2fc collapse expression

leofeyer commented:
@fritzmg Can you please add a unit test that verifies the new behavior?

leofeyer committed:
Allow passing an array of IDs to User::isMemberOf() (see #2834)

Description
-----------

| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2753
| Docs PR or issue | -

To fix #2753 (do not show protected pages in the sitemap if there is no front end user or the user groups do not match), I have adjusted the `User::isMemberOf()` method to also accept an array of group IDs. This allows us to greatly simplify other checks as well:

### Current

```php
$blnFeUserLoggedIn = System::getContainer()->get('contao.security.token_checker')->hasFrontendUser();

while ($objCalendar->next())
{
	if ($objCalendar->protected)
	{
		if (!$blnFeUserLoggedIn || !\is_array($this->User->groups))
		{
			continue;
		}

		$groups = StringUtil::deserialize($objCalendar->groups);

		if (empty($groups) || !\is_array($groups) || \count(array_intersect($groups, $this->User->groups)) < 1)
		{
			continue;
		}
	}

	// …
}
```

### New

```php
$user = null;

if (System::getContainer()->get('contao.security.token_checker')->hasFrontendUser())
{
	$user = FrontendUser::getInstance();
}

while ($objCalendar->next())
{
	if ($objCalendar->protected && (!$user || !$user->isMemberOf(StringUtil::deserialize($objCalendar->groups))))
	{
		continue;
	}

	// …
}
```

Commits
-------

9a924c38 Allow passing an array of IDs to User::isMemberOf()

fritzmg commented:
Yes, I'll do that right away.

fritzmg committed:
add some unit tests

fritzmg committed:
Merge branch '4.11' into fix-content-composition-listener-if-icon-is-null

leofeyer commented:
Thank you @fritzmg.
